public_dropin_notebook_environments/python39_notebook_gpu/agent/cgroup_watchers.py

```diff
@@ -0,0 +1,163 @@
+        cgroup_cpuacct_dir = self._cgroup_mount_dir(subsystem="cpuacct")
+
+        self._memory_usage_file = cgroup_memory_dir / "memory.usage_in_bytes"
+        self._memory_limit_file = cgroup_memory_dir / "memory.limit_in_bytes"
+
+        self._cpu_period_file = cgroup_cpu_dir / "cpu.cfs_period_us"
+        self._cpu_quota_file = cgroup_cpu_dir / "cpu.cfs_quota_us"
+
+        self._cpuacct_usage_file = cgroup_cpuacct_dir / "cpuacct.usage"
+
+    def memory_usage_in_bytes(self) -> int:
+        return self._read_metric(self._memory_usage_file)
+
+    def memory_limit_in_bytes(self) -> int:
+        return self._read_metric(self._memory_limit_file)
+
+    def cpu_quota_micros(self) -> int:
+        return self._read_metric(self._cpu_quota_file)
+
+    def cpu_period_micros(self) -> int:
+        return self._read_metric(self._cpu_period_file)
+
+    def cpuacct_usage_nanos(self) -> int:
+        return self._read_metric(self._cpuacct_usage_file)
+
+    def _read_metric(self, filename: Path) -> int:
+        with open(filename) as f:
+            return int(f.read())
+
+    def _cgroup_mount_dir(self, subsystem: str) -> Path:
+        """
+        :param subsystem: cgroup subsystem like memory, cpu etc.
+        :return: directory where subsystem is mounted
+        """
+        try:
+            with open("/proc/mounts", "r") as f:
+                for line in f.readlines():
+                    split_line = re.split(r"\s+", line)
+                    mount_dir = split_line[1]
+
+                    if "cgroup" in mount_dir:
+                        dirname = mount_dir.split("/")[-1]
+                        subsystems = dirname.split(",")
+
+                        if subsystem in subsystems:
+                            return Path(mount_dir)
+        except FileNotFoundError:
+            ...
+
+        raise CGroupVersionUnsupported
+
+
+class BaseWatcher:
+    def cpu_usage_percentage(self) -> float:
+        raise NotImplementedError
+
+    def memory_usage_percentage(self) -> float:
+        raise NotImplementedError
+
+
+class CGroupWatcher(BaseWatcher):
+    def __init__(self, cgroup_file_reader: CGroupFileReader, system_watcher: SystemWatcher) -> None:
+        self._cgroup_file_reader = cgroup_file_reader
+        self._system_watcher = system_watcher
+
+        self._last_cpu_usage_ts_nanos = 0.0
+        self._last_cpu_cum_usage_nanos = 0.0
+
+    def memory_usage_in_bytes(self) -> float:
+        return self._cgroup_file_reader.memory_usage_in_bytes()
+
+    def memory_limit_in_bytes(self) -> float:
+        cgroup_mem_limit = self._cgroup_file_reader.memory_limit_in_bytes()
+        total_virtual_memory = self._system_watcher.virtual_memory().total
+        return min(cgroup_mem_limit, total_virtual_memory)
+
+    def memory_usage_percentage(self) -> float:
+        return round(self.memory_usage_in_bytes() / self.memory_limit_in_bytes() * 100, 2)
+
+    def cpu_usage_limit_in_cores(self) -> float:
```

Comment:
Where is cpu/mem usage going to be used/exposed?

Reply:
It’s the same code that we have for regular images